[Microsoft Defender Intel Synchronizer] Enhance connector, fix several issues

blauwers opened the pull request:
### Proposed changes

* Add recommended actions, RBAC group names, and educate URL to configuration.
* Increase API request timeout and retry strategy for better resilience.
* Implement logging for failed indicators during import.
* Update expiration datetime logic to consider valid_until field.

### Related issues

* Connector incorrectly handles indicator expiration times.

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

Due to an issue with the OpenCTIApiClient.query() method, I was unable to filter indicators for future valid_until date. Adding this filter causes the query to time out.

romain-filigran committed:
https://github.com/OpenCTI-Platform/connectors/issues/4075

romain-filigran committed:
fix typo

romain-filigran committed:
https://github.com/OpenCTI-Platform/connectors/issues/4077

blauwers committed:
Enhance Microsoft Defender connector:
- Add recommended actions, RBAC group names, and educate URL to configuration.
- Increase API request timeout and retry strategy for better resilience.
- Implement logging for failed indicators during import.
- Update expiration datetime logic to consider valid_until field.

blauwers committed:
Merge branch 'issue/4075' into patch2

blauwers commented:
Supersedes #4076
Related to #4076

blauwers committed:
Refactor Defender connector for performant indicator sync, normalization, and error handling

- Introduced robust indicator_value() sanitization and normalization logic (refangs, strips noise, encodes URLs)
- Replaced legacy .list() with batched GraphQL pagination for large TAXII collections
- Optimized matching logic with dict/set lookups and LRU caching for performance (15k+ observables)
- Added graceful SIGINT shutdown and improved logging/debug visibility
- Set default Defender action to 'Audit' and documented supported values
- Hardened error handling around malformed indicators and API edge cases

Powlinett reviewed (dismissed):
Hi @blauwers ,

thank you for your PR.

I was able to run the connector, it seems to work as intended.

I left some comments, mostly about the config. I'd be glad to help you if anything.

Looking forward your answer 👋

blauwers commented on stream/microsoft-defender-intel-synchronizer/README.md:
```diff
@@ -59,14 +59,17 @@ Below are the parameters you'll need to set for Sentinel Connector:
 | Resource Url Path                                 | `resource_path`     | `MICROSOFT_DEFENDER_INTEL_SYNCHRONIZER_RESOURCE_PATH`           | /        | No        | `/api/indicators`                          | The request URL that will be used which is `/api/indicators`                                                                                                                                  ...
 | Expire Time                                       | `expire_time`       | `MICROSOFT_DEFENDER_INTEL_SYNCHRONIZER_EXPIRE_TIME`             | /        | Yes       | `30`                                       | Number of days for your indicator to expire in Sentinel. Suggestion of `30` as a default                                                                                                      ...
-| Action                                            | `action`            | `MICROSOFT_DEFENDER_INTEL_SYNCHRONIZER_ACTION`                  | /        | No        | `alert`                                    | The action to apply if the indicator is matched from within the targetProduct security tool. Possible values are: `Audit`, `Block`, `Alert`, `Warn`.                                          ...
+| Action                                            | `action`            | `MICROSOFT_DEFENDER_INTEL_SYNCHRONIZER_ACTION`                  | /        | No        | `Audit`                                    | The action to apply if the indicator is matched from within the targetProduct security tool. Possible values are: `Allowed`, `Audit`, `Block`, `BlockAndRemediate`, `Warn`.                   ...
```
Moving from Alert to Audit as the default leaves the old behavior intact. The practical expression by Defender for both actions is the same. (see https://learn.microsoft.com/en-us/defender-endpoint/api/ti-indicator#indicator-types). I am unclear how this change breaks already deployed instances.

blauwers commented on stream/microsoft-defender-intel-synchronizer/src/microsoft_defender_intel_synchronizer_connector/connector.py:
```diff
@@ -120,85 +252,117 @@ def run(self) -> None:
                         and "filters" in result["data"]["taxiiCollection"]
```
Fixed in patch.

blauwers commented on stream/microsoft-defender-intel-synchronizer/src/microsoft_defender_intel_synchronizer_connector/connector.py:
```diff
@@ -85,24 +93,148 @@ def _convert_indicator_to_observables(self, data) -> list[dict]:
+                    col: i for i, col in enumerate(self.config.taxii_collections)
+                }
+
                 # Get OpenCTI indicators
```
Thank you for highlighting this, @Powlinett! I added a clear note to the README explaining that the connector user must have the “Access data sharing -> Manage data sharing” capability, or the connector will fail with a FORBIDDEN_ACCESS error. I also improved the error logging so it’s much more obvious when this permission is missing. Your attention to detail really helps make the connector better for everyone—thanks again!

blauwers committed:
Improve error handling and documentation:
- Require rbac_group_names to be a valid JSON array; terminate connector if invalid
- Add clear README note about required "Manage data sharing" permission
- Enhance error logging for missing permissions and invalid config
- Clarify config variable documentation and maintain consistency

blauwers committed:
Merge branch 'master' into patch2

blauwers commented:
@Powlinett Thank you for the review. I hope that with this patch, this is now ready to merge.

Powlinett commented on stream/microsoft-defender-intel-synchronizer/src/microsoft_defender_intel_synchronizer_connector/connector.py:
```diff
@@ -284,155 +298,16 @@ def handle_sigint(signum, frame):
-                        )
-                        if observable_id not in defender_external_ids_set:
-                            opencti_indicators_to_create.append(observable_data)
+                # ...existing code...
```
Is it intended or a commit mistake?

blauwers replied:
This is entirely a mistake. I am so sorry. Patch forthcoming.

Powlinett reviewed (commented):
Hi @blauwers, thank you for the fixes and your responsiveness 🙏

Can you confirm/fix my last comment please (about the removed existing code)?

Then I guess we'll be ready to merge 🎉

blauwers committed:
Undo editing mistake

Powlinett reviewed (approved):
I approve the PR but please fix the tiny typo first so we can merge it after 😇

Thanks again for your work 🙏

helene-nguyen committed:
[mdis] Update logger
